# Annotation of Microbial Genomes by Microbiome Association (AMGMA)

The purpose of this repository is to allow users to annotate a collection of
microbial genomes on the basis of their association with microbiome survey
datasets. 

### Background

Diving more deeply, every microbial genome can be viewed as consisting of a
collection of genetic components, such as genes. In turn, each of those genetic
components (e.g. genes) can be associated with some outcome of interest via
the analysis of microbiome surveys by metagenomic sequencing. The tool most easily
used for this purpose (and compatible for downstream analysis by AMGMA) is
[GeneShot](https://www.github.com/Golob-Minot/GeneShot).

After each microbial gene has been associated with outcome of interest (such
as a particular human disease), then those association metrics can be summarized
over a set of microbial genomes by:

  1. Aligning the reference genes against each genome's genes, and
  2. Calculating summary metrics for each genome, such as:
    * Proportion of genes passing FDR filter for association
    * Average estimated coefficient of association for those genes
  3. Formatting a genome 'map' showing the relative physical location for those genes throughout the genome

### Running AMGMA

_Formatting a Reference Genome Database_:

Using a collection of reference genomes, run the `AMGMA/build_db` script.

_Annotate Reference Genomes_:

Using that genome database, annotate with a set of microbiome association results
generated by `geneshot` with the main `AMGMA` executable.
