# Annotation of Microbial Genomes by Microbiome Association (AMGMA)

The purpose of this repository is to allow users to annotate a collection of
microbial genomes on the basis of their association with microbiome survey
datasets. 

### Background

Diving more deeply, every microbial genome can be viewed as consisting of a
collection of genetic components, such as genes. In turn, each of those genetic
components (e.g. genes) can be associated with some outcome of interest via
the analysis of microbiome surveys by metagenomic sequencing. The tool most easily
used for this purpose (and compatible for downstream analysis by AMGMA) is
[GeneShot](https://www.github.com/Golob-Minot/GeneShot).

After each microbial gene has been associated with outcome of interest (such
as a particular human disease), then those association metrics can be summarized
over a set of microbial genomes by:

  1. Aligning the reference genes against each genome's genes, and
  2. Calculating summary metrics for each genome, such as:
    * Proportion of genes passing FDR filter for association
    * Average estimated coefficient of association for those genes
  3. Formatting a genome 'map' showing the relative physical location for those genes throughout the genome

### Running AMGMA

_Formatting a Reference Genome Database_:

Using a collection of reference genomes, run the `AMGMA/build_db` script.

Input data for this step is simply a collection of microbial genomes, and
a manifest CSV file listing those genomes. The CSV must have the following
columns: `uri,id,name`:

  * `uri`: Location of the genome FASTA
  * `id`: Unique alphanumeric ID for each genome
  * `name`: Longer description of each genome, with whitespaces allowed (but no commas)

```
Usage:

nextflow run FredHutch/AMGMA/build_db.nf <ARGUMENTS>

Required Arguments:
  --manifest            CSV file listing samples (see below)
  --output_folder       Folder to write output files to
  --output_prefix       Prefix to use for output file names

Optional Arguments:
  --min_identity        Percent identity threshold used to pick centroids (default: 90)
  --min_coverage        Percent coverage threshold used to pick centroids (default: 50)

Manifest:
  The manifest is a CSV listing all of the genomes to be used for the database.
  The manifest much contain the column headers: uri,id,name
  The URI may start with ftp://, s3://, or even just be a path to a local file.
  The ID must be unique, and only contain a-z, A-Z, 0-9, or _.
  The NAME may be longer and contain whitespaces, but may not contain a comma.
```

_Annotate Reference Genomes_:

Using that genome database, annotate with a set of microbiome association results
generated by `geneshot` with the main `AMGMA` executable.

```
Usage:

nextflow run FredHutch/AMGMA <ARGUMENTS>

Required Arguments:
  --geneshot_hdf        Results HDF file output by GeneShot, containing CAG information
  --geneshot_fasta      Gzip-compressed FASTA with the gene catalog generated by GeneShot
  --db_dmnd             Microbial genome database, DMND format
  --db_hdf              Microbial genome database, HDF format
  --output_folder       Folder to write output HDF file within
  --output_hdf          Name of the output HDF file to write to the output folder

Optional Arguments:
  --min_coverage        Minimum coverage required for alignment (default: 50)
  --min_identity        Minimum percent identity required for alignment (default: 80)
  --batchsize           Number of genomes to process in a given batch
  --fdr_method          Method used for FDR correction (default: fdr_bh)
  --alpha               Alpha value used for FDR correction (default: 0.2)

Database files:
The DMND and HDF files used as inputs for this script are produced by the build_db.nf
function provided as part of this project.

Input HDF:
The information on CAGs used as an input for this analysis is formatted as the output
from the GeneShot pipeline (more information at github.com/Golob-Minot/GeneShot).

Input FASTA:
Another output from the GeneShot pipeline is the set of genes which were identified,
in gzip-compressed FASTA format, specified as input here with --geneshot_fasta

Output HDF:
The output from this pipeline is an HDF file which contains all of the data from the
input HDF, as well as the additional tables,

  * /genomes/summary/<feature>
  * /genomes/deaeil/<feature>/<genome_id>

  for each <feature> tested in the input, and for each <genome_id> in the database
```

### How AMGMA Works

The concept of AMGMA is that it will compare a set of genes from an external
gene catalog against a collection of genomes. Each of those 'catalog genes'
has been associated with some measure(s) of interest using the intermediate
entity of Co-Abundant Gene Groups (CAGs). Each CAG has been analyzed for its
estimated coefficient of association with some measure of interest, and we
also have a p-value for that association. At the level of the CAGs, we can
apply some False Discovery Rate (FDR) correction, and then propagate those
corrected p-values and estimated coefficients to each of the catalog genes
found within each CAG.

The next step is to figure out which of the 'catalog genes' is present in each
of the reference genomes, along with the position of those genes in the genome.
We do this by:
  1. Identifying all of the protein-coding regions in each genome, 'genome genes'
  2. Clustering those 'genome genes' by percent identity making 'genome centroids'
  3. Aligning the 'catalog genes' against the 'genome centroids'
  4. Summarizing each genome on the basis of the estimated coefficient and 
  FDR-corrected p-value for each 'catalog gene', going through the intermediate
  step of 'genome centroids' and finally to 'genome genes.'

### Database Format

The reference database made with the `build_db` script will create two files,
a DMND database of 'genome centroids', and an HDF store with a few tables:

  * `/manifest`: The manifest used to build the database
  * `/genome_centroids`: A table linking each `genome_gene` to a `genome_centroid`
  * `/genomes/<id>`: A table with the `genome_gene` for each genome, by `id`


### Output Format

After running the main `AMGMA` script, two output files will be created, one
with the full results, and one with a summary. These will both be HDF files
and will include everything present in the HDF file used as an input (matching
the output from `geneshot`). The idea is that `geneshot` was run in such a way
as to include some statistical analysis testing for the association of every
individual CAG with some features of interest. And so there may be estimated
coefficients of association for multiple `<feature>` elements in the HDF file
used as input for `AMGMA`, whose output will then include the additional tables:

  * `/genomes/summary/<feature>`: A table with summary metrics for each genome
  * `/genomes/detail/<feature>/<id>`: For each genome, a map of gene locations
